src/flamenco/vm/syscall/fd_vm_syscall_util.c

```diff
@@ -669,58 +667,72 @@ fd_vm_syscall_sol_memset( /**/            void *  _vm,
   fd_vm_t * vm = (fd_vm_t *)_vm;
+  *_ret = 0;
 
   /* https://github.com/anza-xyz/agave/blob/master/programs/bpf_loader/src/syscalls/mem_ops.rs#L115 */
 
   FD_VM_CU_MEM_OP_UPDATE( vm, sz );
 
-  ulong FD_FN_UNUSED dst_region = dst_vaddr >> 32;
+  if( FD_UNLIKELY( !sz ) ) {
+    return FD_VM_SUCCESS;
+  }
+
+  ulong   dst_region = fd_ulong_min( dst_vaddr >> 32, 5UL );
+  ulong   dst_offset = dst_vaddr & 0xffffffffUL;
+  uchar * dst_haddr;
+
   int b = (int)(c & 255UL);
 
-  if( dst_region!=4UL || !FD_FEATURE_ACTIVE( vm->instr_ctx->slot_ctx, bpf_account_data_direct_mapping ) ) {
-    void * dst = FD_VM_MEM_SLICE_HADDR_ST( vm, dst_vaddr, 1UL, sz );
-    fd_memset( dst, b, sz );
-  } else {
-    /* Syscall manages the pointer accesses directly and will report in the 
-       case of bad memory accesses. This syscall doesn't have the same nuance
-       as what is described in memcmp, because the VM will abort as soon as
-       out of bounds memory tries to get written to. Therefore, we don't need
-       to fault early. */
-    ulong sz_left              = sz;
-    ulong dst_offset           = dst_vaddr & 0xffffffffUL;
-    ulong region_idx           = fd_vm_get_input_mem_region_idx( vm, dst_offset );
-    ulong region_offset        = dst_offset - vm->input_mem_regions[region_idx].vaddr_offset;
-    ulong bytes_left_in_region = fd_ulong_sat_sub(vm->input_mem_regions[region_idx].region_sz, region_offset);
-    uchar * haddr              = (uchar*)(vm->input_mem_regions[region_idx].haddr + region_offset);
-
-    if( FD_UNLIKELY( !bytes_left_in_region ) ) {
-      *_ret = 1UL;
-      return FD_VM_ERR_INVAL;
+  if( !vm->direct_mapping ) {
+    dst_haddr = FD_VM_MEM_HADDR_ST( vm, dst_vaddr, FD_VM_ALIGN_RUST_U8, sz );
+    fd_memset( dst_haddr, b, sz );
+  } else if( dst_region!=4UL ) {
+    /* I acknowledge that this is not an ideal implementation, but Agave will
+       memset as many bytes as possible until it reaches an unwritable section.
+       This is purely just for fuzzing conformance, sigh... */
+    dst_haddr = (uchar*)FD_VM_MEM_HADDR_ST_FAST( vm, dst_vaddr );
+    ulong dst_bytes_rem_in_cur_region = fd_ulong_sat_sub( vm->region_st_sz[ dst_region ], dst_offset );
+    ulong bytes_to_set                = fd_ulong_min( sz, dst_bytes_rem_in_cur_region );
+    fd_memset( dst_haddr, b, bytes_to_set );
+    if( FD_UNLIKELY( bytes_to_set<sz ) ) {
+      FD_VM_ERR_FOR_LOG_EBPF( vm, FD_VM_ERR_EBPF_ACCESS_VIOLATION );
+      return FD_VM_ERR_SIGSEGV;
     }
-
-    while( sz_left ) {
-      if( FD_UNLIKELY( region_idx>=vm->input_mem_regions_cnt ) ) {
-        *_ret = 1UL;
-        return FD_VM_ERR_INVAL;
-      }
-      if( FD_UNLIKELY( !vm->input_mem_regions[region_idx].is_writable ) ) {
-        *_ret = 1UL;
-        return FD_VM_ERR_INVAL;
+  } else {
+    /* In this case, we are in the input region AND direct mapping is enabled.
+       Get the haddr and input region and check if it's writable. */
```

Comment:
id rather not tbh because it would just make the while loop more complicated to read since behavior to handle input vs non-input regions are vastly different besides the actual memsetting